pwa-migration-explainer.md

```diff
@@ -0,0 +1,199 @@
+# PWA Origin Migration
+
+## Authors:
+
+*   Marijn Kruisselbrink (Google)
+
+## Participate
+
+*   **Issue Tracker:** [WICG/manifest-incubations on GitHub](https://github.com/WICG/manifest-incubations/issues) (for spec-related issues)
+*   **Chromium Implementation Bug:** [crbug.com/380486617](https://issues.chromium.org/380486617) (for implementation-specific discussion)
+
+## Introduction
+
+When a user installs a Progressive Web App (PWA), its identity and security context are tightly bound to its web origin (e.g., `app.example.com`). This presents a significant challenge for developers who need to change their PWA's origin due to rebranding, domain restructuring, or technical re-architecture. Currently, such a change forces users to manually uninstall the old app and reinstall the new one, leading to a disruptive experience and potential user loss. This proposal introduces a mechanism for developers to seamlessly migrate an installed PWA to a new, same-site origin, preserving user trust and permissions.
+
+## User-Facing Problem
+
+Imagine you have the "SocialApp" app installed on your computer from `www.example.com/social`. One day, the company decides to move the app to its own dedicated home at `social.example.com`. Without a migration mechanism, the app you have installed would either break or redirect you to the new site in a generic browser window, losing its app-like feel. You would have to figure out that you need to uninstall the old app and install the new one from the new address. You might also lose your settings, like whether you've allowed the app to send notifications.
+
+This feature aims to make that transition seamless. Instead of a broken experience, the app would notify you of an available update. With your approval, the app would relaunch from its new home at `social.example.com`, with your notification settings intact.
+
+### Goals
+
+*   **Seamless User Experience:** Users should be able to transition to a new PWA origin with a simple, non-disruptive update flow.
+*   **Developer Enablement:** Provide a clear mechanism for developers to migrate their PWAs for reasons like rebranding (e.g., "ConnectApp" to "ConnectX"), SEO optimization (e.g., `photo-app.example.com` to `photos.example.com/app`), or backend architecture changes (e.g., `mail.example.com` to `app.example.com`).
+*   **Permission Persistence:** User-granted permissions (e.g., notifications, location) should be securely and transparently migrated to the new origin where possible.
+*   **Security:** The migration process must be secure, preventing malicious takeovers or phishing attacks.
+
+### Non-goals
+
+*   **Cross-Site Migration:** This feature will not support migration between origins that are not "same-site" (e.g., from `example.com` to `another-site.com`).
```

Comment:
Is it worth saying why this is a non-goal? The arguably biggest PWA migration (`twitter.com` → `x.com`) would not have profited from this proposal had it existed at the time.

Reply:
Added a bit more context and clarified why not and that we might want to explore this in the future